# 2014222 - Semester 2 CA-01 -April 2024¶

# Part 2

# Step 1: Import Python Libraries

In [15]:
# Utilise Python programming language in order to comply with the requisites of the assessment and perform adequate Machine
# Learning algorithms to discover and deliver insights.

# Import the necessary libraries (Pandas) in order to perform data cleansing. These are the libraries that are 
# conventionally used as a common practice in order to perform mathematical and statistical operations during a data analysis
# process

import pandas as pd

# Import the Plotly library in order to perform data visualisation procedures

import plotly.express as px
import plotly.graph_objects as go

# I am using this line of code to see all columns in a wide DataFrame

pd.set_option('display.max_columns', None)

# Step 2: Reading Dataset

In [16]:
# Read the CSV file by applying the pd.read_csv() function.

nba_data = pd.read_csv('NBA Dataset CSV C1.csv')

Reference to the dataset:

https://stats.nba.com/stats/leagueLeaders?LeagueID=00&PerMode=Totals&Scope=S&Season=2012-13&SeasonType=Regular%20Season&StatCategory=PTS

# Analyzing the Data

In [17]:
# Print the NBA table data

nba_data

Year       Season_type  PLAYER_ID  RANK           PLAYER     TEAM_ID  \
0     2012-13  Regular%20Season     201142     1     Kevin Durant  1610612760   
1     2012-13  Regular%20Season        977     2      Kobe Bryant  1610612747   
2     2012-13  Regular%20Season       2544     3     LeBron James  1610612748   
3     2012-13  Regular%20Season     201935     4     James Harden  1610612745   
4     2012-13  Regular%20Season       2546     5  Carmelo Anthony  1610612752   
...       ...               ...        ...   ...              ...         ...   
7288  2021-22          Playoffs    1629006   206      Josh Okogie  1610612750   
7289  2021-22          Playoffs    1630556   206  Kessler Edwards  1610612751   
7290  2021-22          Playoffs    1630201   206    Malachi Flynn  1610612761   
7291  2021-22          Playoffs     202693   206  Markieff Morris  1610612748   
7292  2021-22          Playoffs     200794   206     Paul Millsap  1610612755   

     TEAM  GP   MIN  FGM   FGA  FG_PCT  FG3M  FG3A  FG3_PCT  FTM  FTA  FT_PCT  \
0     OKC  81  3119  731  1433   0.510   139   334    0.416  679  750   0.905   
1     LAL  78  3013  738  1595   0.463   132   407    0.324  525  626   0.839   
2     MIA  76  2877  765  1354   0.565   103   254    0.406  403  535   0.753   
3     HOU  78  2985  585  1337   0.438   179   486    0.368  674  792   0.851   
4     NYK  67  2482  669  1489   0.449   157   414    0.379  425  512   0.830   
...   ...  ..   ...  ...   ...     ...   ...   ...      ...  ...  ...     ...   
7288  MIN   1     2    0     0   0.000     0     0    0.000    0    0   0.000   
7289  BKN   2     7    0     0   0.000     0     0    0.000    0    0   0.000   
7290  TOR   6    36    0     7   0.000     0     3    0.000    0    0   0.000   
7291  MIA   2     3    0     1   0.000     0     0    0.000    0    0   0.000   
7292  PHI   1     6    0     0   0.000     0     0    0.000    0    0   0.000   

      OREB  DREB  REB  AST  STL  BLK  TOV   PF   PTS   EFF  AST_TOV  STL_TOV  
0       46   594  640  374  116  105  280  143  2280  2462     1.34     0.41  
1       66   367  433  469  106   25  287  173  2133  1921     1.63     0.37  
2       97   513  610  551  129   67  226  110  2036  2446     2.44     0.57  
3       62   317  379  455  142   38  295  178  2023  1872     1.54     0.48  
4      134   326  460  171   52   32  175  205  1920  1553     0.98     0.30  
...    ...   ...  ...  ...  ...  ...  ...  ...   ...   ...      ...      ...  
7288     0     0    0    0    0    0    0    0     0     0     0.00     0.00  
7289     0     0    0    1    1    0    1    3     0     1     1.00     1.00  
7290     2     1    3    3    1    0    1    6     0    -1     3.00     1.00  
7291     0     1    1    0    0    0    1    2     0    -1     0.00     0.00  
7292     0     1    1    1    0    0    0    1     0     2     0.00     0.00  

[7293 rows x 30 columns]

In [3]:
# Print the dimensions of the NBA DataFrame.

nba_data.shape

(7293, 30)

As we can see that the dataframe contains In this case is NBA 7293 rows × 30 columns

In [4]:
# Print the first 5 rows of the NBA dataframe by applying the .head() method,
# This method is will display the top 5 observations of the dataset

nba_data.head()

Year       Season_type  PLAYER_ID  RANK           PLAYER     TEAM_ID  \
0  2012-13  Regular%20Season     201142     1     Kevin Durant  1610612760   
1  2012-13  Regular%20Season        977     2      Kobe Bryant  1610612747   
2  2012-13  Regular%20Season       2544     3     LeBron James  1610612748   
3  2012-13  Regular%20Season     201935     4     James Harden  1610612745   
4  2012-13  Regular%20Season       2546     5  Carmelo Anthony  1610612752   

  TEAM  GP   MIN  FGM   FGA  FG_PCT  FG3M  FG3A  FG3_PCT  FTM  FTA  FT_PCT  \
0  OKC  81  3119  731  1433   0.510   139   334    0.416  679  750   0.905   
1  LAL  78  3013  738  1595   0.463   132   407    0.324  525  626   0.839   
2  MIA  76  2877  765  1354   0.565   103   254    0.406  403  535   0.753   
3  HOU  78  2985  585  1337   0.438   179   486    0.368  674  792   0.851   
4  NYK  67  2482  669  1489   0.449   157   414    0.379  425  512   0.830   

   OREB  DREB  REB  AST  STL  BLK  TOV   PF   PTS   EFF  AST_TOV  STL_TOV  
0    46   594  640  374  116  105  280  143  2280  2462     1.34     0.41  
1    66   367  433  469  106   25  287  173  2133  1921     1.63     0.37  
2    97   513  610  551  129   67  226  110  2036  2446     2.44     0.57  
3    62   317  379  455  142   38  295  178  2023  1872     1.54     0.48  
4   134   326  460  171   52   32  175  205  1920  1553     0.98     0.30

In [5]:
# Print the last 5 rows of the NBA dataframe by applying the .tail() method,
# Will display the last 5 observations of the dataset

nba_data.tail()

Year Season_type  PLAYER_ID  RANK           PLAYER     TEAM_ID TEAM  \
7288  2021-22    Playoffs    1629006   206      Josh Okogie  1610612750  MIN   
7289  2021-22    Playoffs    1630556   206  Kessler Edwards  1610612751  BKN   
7290  2021-22    Playoffs    1630201   206    Malachi Flynn  1610612761  TOR   
7291  2021-22    Playoffs     202693   206  Markieff Morris  1610612748  MIA   
7292  2021-22    Playoffs     200794   206     Paul Millsap  1610612755  PHI   

      GP  MIN  FGM  FGA  FG_PCT  FG3M  FG3A  FG3_PCT  FTM  FTA  FT_PCT  OREB  \
7288   1    2    0    0     0.0     0     0      0.0    0    0     0.0     0   
7289   2    7    0    0     0.0     0     0      0.0    0    0     0.0     0   
7290   6   36    0    7     0.0     0     3      0.0    0    0     0.0     2   
7291   2    3    0    1     0.0     0     0      0.0    0    0     0.0     0   
7292   1    6    0    0     0.0     0     0      0.0    0    0     0.0     0   

      DREB  REB  AST  STL  BLK  TOV  PF  PTS  EFF  AST_TOV  STL_TOV  
7288     0    0    0    0    0    0   0    0    0      0.0      0.0  
7289     0    0    1    1    0    1   3    0    1      1.0      1.0  
7290     1    3    3    1    0    1   6    0   -1      3.0      1.0  
7291     1    1    0    0    0    1   2    0   -1      0.0      0.0  
7292     1    1    1    0    0    0   1    0    2      0.0      0.0

In [6]:
# Getting information about the NBA dataframe by applying the .info() method,
# Will display number of records in each column, data having null or not null, Data type,
# memory usage of the dataset
# helps to understand the data type and information about data

nba_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7293 entries, 0 to 7292
Data columns (total 30 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Year         7293 non-null   object 
 1   Season_type  7293 non-null   object 
 2   PLAYER_ID    7293 non-null   int64  
 3   RANK         7293 non-null   int64  
 4   PLAYER       7293 non-null   object 
 5   TEAM_ID      7293 non-null   int64  
 6   TEAM         7293 non-null   object 
 7   GP           7293 non-null   int64  
 8   MIN          7293 non-null   int64  
 9   FGM          7293 non-null   int64  
 10  FGA          7293 non-null   int64  
 11  FG_PCT       7293 non-null   float64
 12  FG3M         7293 non-null   int64  
 13  FG3A         7293 non-null   int64  
 14  FG3_PCT      7293 non-null   float64
 15  FTM          7293 non-null   int64  
 16  FTA          7293 non-null   int64  
 17  FT_PCT       7293 non-null   float64
 18  OREB         7293 non-null   int64  
 19  DREB  

# Check for Duplication

In [7]:
# Getting several unique values in each columnan the data description the NBA dataframe
# by applying the .nunique() method,
# Will display continuous and categorical columns in the data.
# Duplicated data can be handled or removed based on further analysis
# helps to understand the data type and information about data

nba_data.nunique()

Year             10
Season_type       2
PLAYER_ID      1386
RANK            537
PLAYER         1382
TEAM_ID          30
TEAM             31
GP               83
MIN            2331
FGM             670
FGA            1210
FG_PCT          483
FG3M            261
FG3A            566
FG3_PCT         379
FTM             428
FTA             517
FT_PCT          539
OREB            285
DREB            590
REB             748
AST             544
STL             165
BLK             175
TOV             289
PF              276
PTS            1449
EFF            1567
AST_TOV         445
STL_TOV         226
dtype: int64

In [8]:
# Identifying null values in the NBA dataframe by applying the .isnull() method,

nba_data.isnull()

Year  Season_type  PLAYER_ID   RANK  PLAYER  TEAM_ID   TEAM     GP  \
0     False        False      False  False   False    False  False  False   
1     False        False      False  False   False    False  False  False   
2     False        False      False  False   False    False  False  False   
3     False        False      False  False   False    False  False  False   
4     False        False      False  False   False    False  False  False   
...     ...          ...        ...    ...     ...      ...    ...    ...   
7288  False        False      False  False   False    False  False  False   
7289  False        False      False  False   False    False  False  False   
7290  False        False      False  False   False    False  False  False   
7291  False        False      False  False   False    False  False  False   
7292  False        False      False  False   False    False  False  False   

        MIN    FGM    FGA  FG_PCT   FG3M   FG3A  FG3_PCT    FTM    FTA  \
0     False  False  False   False  False  False    False  False  False   
1     False  False  False   False  False  False    False  False  False   
2     False  False  False   False  False  False    False  False  False   
3     False  False  False   False  False  False    False  False  False   
4     False  False  False   False  False  False    False  False  False   
...     ...    ...    ...     ...    ...    ...      ...    ...    ...   
7288  False  False  False   False  False  False    False  False  False   
7289  False  False  False   False  False  False    False  False  False   
7290  False  False  False   False  False  False    False  False  False   
7291  False  False  False   False  False  False    False  False  False   
7292  False  False  False   False  False  False    False  False  False   

      FT_PCT   OREB   DREB    REB    AST    STL    BLK    TOV     PF    PTS  \
0      False  False  False  False  False  False  False  False  False  False   
1      False  False  False  False  False  False  False  False  False  False   
2      False  False  False  False  False  False  False  False  False  False   
3      False  False  False  False  False  False  False  False  False  False   
4      False  False  False  False  False  False  False  False  False  False   
...      ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
7288   False  False  False  False  False  False  False  False  False  False   
7289   False  False  False  False  False  False  False  False  False  False   
7290   False  False  False  False  False  False  False  False  False  False   
7291   False  False  False  False  False  False  False  False  False  False   
7292   False  False  False  False  False  False  False  False  False  False   

        EFF  AST_TOV  STL_TOV  
0     False    False    False  
1     False    False    False  
2     False    False    False  
3     False    False    False  
4     False    False    False  
...     ...      ...      ...  
7288  False    False    False  
7289  False    False    False  
7290  False    False    False  
7291  False    False    False  
7292  False    False    False  

[7293 rows x 30 columns]

As we can see the NBA dataframe null values are false, meaning there is not missing values

In [9]:
# Identifying missing values in the NBA dataframe by applying the .isna().sum() methods,
# I am using this to get the number of missing records in each column

nba_data.isnull().sum()

Year           0
Season_type    0
PLAYER_ID      0
RANK           0
PLAYER         0
TEAM_ID        0
TEAM           0
GP             0
MIN            0
FGM            0
FGA            0
FG_PCT         0
FG3M           0
FG3A           0
FG3_PCT        0
FTM            0
FTA            0
FT_PCT         0
OREB           0
DREB           0
REB            0
AST            0
STL            0
BLK            0
TOV            0
PF             0
PTS            0
EFF            0
AST_TOV        0
STL_TOV        0
dtype: int64

As we can see the NBA dataframe does not contain any duplicates.

In [10]:
# Identifying the percentage of missing values in the NBA dataframe by applying the .isna().sum() methods,
# I am using this to get the number of missing records in each column

(nba_data.isnull().sum()/(len(nba_data)))* 100

Year           0.0
Season_type    0.0
PLAYER_ID      0.0
RANK           0.0
PLAYER         0.0
TEAM_ID        0.0
TEAM           0.0
GP             0.0
MIN            0.0
FGM            0.0
FGA            0.0
FG_PCT         0.0
FG3M           0.0
FG3A           0.0
FG3_PCT        0.0
FTM            0.0
FTA            0.0
FT_PCT         0.0
OREB           0.0
DREB           0.0
REB            0.0
AST            0.0
STL            0.0
BLK            0.0
TOV            0.0
PF             0.0
PTS            0.0
EFF            0.0
AST_TOV        0.0
STL_TOV        0.0
dtype: float64

As we can see the percentage of missing values for the columns is 0%

In [11]:
# Print the NBA table data. Only 15 rows are used as a sample.

nba_data.sample(15)

Year       Season_type  PLAYER_ID  RANK            PLAYER  \
4986  2019-20  Regular%20Season     201568    41  Danilo Gallinari   
3529  2017-18  Regular%20Season     201959    76        Taj Gibson   
5472  2019-20  Regular%20Season    1629716   525    Marques Bolden   
4024  2017-18          Playoffs    1627732    31       Ben Simmons   
1986  2014-15          Playoffs     203082   133     Terrence Ross   
5661  2019-20          Playoffs    1627788   186    Furkan Korkmaz   
7185  2021-22          Playoffs    1630631   110     Jose Alvarado   
1490  2014-15  Regular%20Season       2449   129        Luis Scola   
5599  2019-20          Playoffs     202355   125  Hassan Whiteside   
2751  2015-16          Playoffs     203143   207    Pablo Prigioni   
3804  2017-18  Regular%20Season    1626143   351     Jahlil Okafor   
1312  2013-14          Playoffs       2422   154      John Salmons   
3939  2017-18  Regular%20Season     203481   486       Jeff Withey   
4686  2018-19  Regular%20Season     203112   483        Quincy Acy   
4789  2018-19          Playoffs    1628378    55  Donovan Mitchell   

         TEAM_ID TEAM  GP   MIN  FGM  FGA  FG_PCT  FG3M  FG3A  FG3_PCT  FTM  \
4986  1610612760  OKC  62  1834  358  817   0.438   178   439    0.405  266   
3529  1610612750  MIN  82  2726  425  737   0.577     7    35    0.200  142   
5472  1610612739  CLE   1     3    0    0   0.000     0     0    0.000    0   
4024  1610612755  PHI  10   369   61  125   0.488     0     1    0.000   41   
1986  1610612761  TOR   4   107   11   29   0.379     6    18    0.333    0   
5661  1610612755  PHI   4    40    0    7   0.000     0     6    0.000    3   
7185  1610612740  NOP   6   117   16   33   0.485     6    16    0.375   10   
1490  1610612754  IND  81  1659  300  642   0.467     5    20    0.250  158   
5599  1610612757  POR   5   106   13   24   0.542     1     1    1.000    7   
2751  1610612746  LAC   5    26    0    5   0.000     0     2    0.000    0   
3804  1610612751  BKN  28   353   77  138   0.558     1     4    0.250   21   
1312  1610612761  TOR   6    77    5   17   0.294     2    12    0.167    1   
3939  1610612742  DAL  10    39    6   16   0.375     2    10    0.200    1   
4686  1610612756  PHX  10   123    4   18   0.222     2    15    0.133    7   
4789  1610612762  UTA   5   193   36  112   0.321    11    43    0.256   24   

      FTA  FT_PCT  OREB  DREB  REB  AST  STL  BLK  TOV   PF   PTS   EFF  \
4986  298   0.893    34   288  322  119   42    5   76   75  1160  1081   
3529  185   0.768   194   390  584   99   62   59   89  218   999  1359   
5472    0   0.000     0     2    2    0    1    0    0    1     0     3   
4024   58   0.707    16    78   94   77   17    8   44   34   163   234   
1986    0   0.000     0     6    6    4    3    4    2   10    28    25   
5661    5   0.600     0     6    6    2    1    0    2    3     3     1   
7185   13   0.769     2     6    8    9    7    1    7   18    48    46   
1490  226   0.699   138   388  526  105   47   18   94  192   763   955   
5599   14   0.500     7    28   35    2    1   10    6   17    34    58   
2751    0   0.000     2     1    3    7    0    0    3    8     0     2   
3804   29   0.724    21    64   85   12    3   18   21   51   176   204   
1312    1   1.000     0     6    6    5    2    0    1    9    13    13   
3939    2   0.500     1     9   10    0    0    3    2    3    15    15   
4686   10   0.700     3    22   25    8    1    4    4   24    17    34   
4789   33   0.727     7    18   25   16    8    1   21   15   107    51   

      AST_TOV  STL_TOV  
4986     1.57     0.55  
3529     1.11     0.70  
5472     0.00     0.00  
4024     1.75     0.39  
1986     2.00     1.50  
5661     1.00     0.50  
7185     1.29     1.00  
1490     1.12     0.50  
5599     0.33     0.17  
2751     2.33     0.00  
3804     0.57     0.14  
1312     5.00     2.00  
3939     0.00     0.00  
4686     2.00     0.25  
4789     0.76     0.38

In [12]:
# Checking if there any missing or null values in the NBA dataframe by applying the .isna() method,

# and the .sum() method to return the sum of the duplicate values in the whole dataframe.

# The duplicated() method would return a series of 'True' or 'False', based on whether a row contains
# a duplicate value or not. Whereas, by adding the .sum() method, which
# is conventionally used to obtain the sum of a series of values,
# it is possible to obtain the whole sum of all the duplicated
# values contained in the dataframe

nba_data.isna().sum()

Year           0
Season_type    0
PLAYER_ID      0
RANK           0
PLAYER         0
TEAM_ID        0
TEAM           0
GP             0
MIN            0
FGM            0
FGA            0
FG_PCT         0
FG3M           0
FG3A           0
FG3_PCT        0
FTM            0
FTA            0
FT_PCT         0
OREB           0
DREB           0
REB            0
AST            0
STL            0
BLK            0
TOV            0
PF             0
PTS            0
EFF            0
AST_TOV        0
STL_TOV        0
dtype: int64

# Step 3: Data Reduction

In [13]:
# Droppin the columns that I am not going to use in the NBA dataframe by applying the .drop method,

# In this case is raking and efficiency columns

nba_data.drop(columns=['RANK','EFF'], inplace=True)